File: Tools/StockStatistic.cs
﻿using System.Globalization;
using System.Collections.Generic;

using Android.App;

namespace VorratsUebersicht
{
    public class StockStatistic
    {
        internal int count = 0;
        internal decimal quantity = 0;
        internal Dictionary<string, decimal> sum_menge = new Dictionary<string, decimal>();
        decimal sum_warnung = 0;
        decimal sum_abgelaufen = 0;
        decimal sum_kcal = 0;
        decimal sum_price = 0;

        internal void AddWarningLevel1(decimal quantity)
        {
            sum_warnung += quantity;
        }

        internal void AddWarningLevel2(decimal quantity)
        {
            sum_abgelaufen += quantity;
        }
        
        private void AddUnitQuantity(string unit, decimal size, decimal quantity)
        {
            if (string.IsNullOrEmpty(unit))
                unit = string.Empty;

            if (!this.sum_menge.ContainsKey(unit))
            {
                this.sum_menge.Add(unit, size * quantity);
            }
            else
            {
                this.sum_menge[unit] += size * quantity;
            }
        }

        private void AddCalorie(decimal quantity, int calorie)
        {
            sum_kcal += quantity * calorie;
        }

        internal void AddStorageItem(StorageItemQuantityResult storegeItem)
        {
            this.count ++;
            this.quantity += storegeItem.Quantity;
            this.AddUnitQuantity(storegeItem.Unit, storegeItem.Size, storegeItem.Quantity);
            this.AddCalorie(storegeItem.Quantity, storegeItem.Calorie);
            this.AddCosts(storegeItem.Quantity, storegeItem.Price);
        }

        private void AddCosts(decimal quantity, decimal? price)
        {
            if (!price.HasValue)
                return;

            this.sum_price += quantity * price.Value;
        }

        internal string GetStatistic(Activity activity)
        {
            string status;
            if (this.count == 1)
                status = string.Format(activity.Resources.GetString(Resource.String.ArticleListSummary_Position), this.count);
            else
                status = string.Format(activity.Resources.GetString(Resource.String.ArticleListSummary_Positions), this.count);

            status += ", " + string.Format(CultureInfo.CurrentUICulture, activity.Resources.GetString(Resource.String.StorageListSummary_Quantity), this.quantity);

            if (this.sum_menge.Count > 0)
            {
                string mengeListe = string.Empty;

                this.ConvertUnits();

                foreach(var menge in this.sum_menge)
                {
                    if (menge.Value == 0)
                        continue;

                    if (!string.IsNullOrEmpty(mengeListe))
                        mengeListe += ", ";
                    mengeListe += string.Format(CultureInfo.CurrentUICulture, "{0:#,0.######} {1}", menge.Value, menge.Key);
                }

                if (!string.IsNullOrEmpty(mengeListe))
                {
                    status += ", " + string.Format(activity.Resources.GetString(Resource.String.StorageListSummary_Amount), mengeListe);
                }
            }

            if (sum_kcal        > 0) status += ", " + string.Format(CultureInfo.CurrentUICulture, activity.Resources.GetString(Resource.String.StorageListSummary_Calories),   sum_kcal);
            if (sum_price       > 0) status += ", " + string.Format(CultureInfo.CurrentUICulture, activity.Resources.GetString(Resource.String.StorageListSummary_Value),   sum_price, CultureInfo.CurrentCulture.NumberFormat.CurrencySymbol);
            if (sum_warnung     > 0) status += ", " + string.Format(CultureInfo.CurrentUICulture, activity.Resources.GetString(Resource.String.StorageListSummary_Warning), sum_warnung);
            if (sum_abgelaufen  > 0) status += ", " + string.Format(CultureInfo.CurrentUICulture, activity.Resources.GetString(Resource.String.StorageListSummary_Off),  sum_abgelaufen);
            
            return status;
        }

        internal void ConvertUnits()
        {
            if (this.sum_menge.ContainsKey("ml") && this.sum_menge.ContainsKey("l"))
            {
                decimal ml = this.sum_menge["ml"];
                decimal l  = this.sum_menge["l"];

                decimal liter = l + ml / 1000;

                this.sum_menge.Remove("ml");
                this.sum_menge["l"] = liter;
            }

            if (this.sum_menge.ContainsKey("cl") && this.sum_menge.ContainsKey("l"))
            {
                decimal cl = this.sum_menge["cl"];
                decimal l  = this.sum_menge["l"];

                decimal liter = l + cl / 100;

                this.sum_menge.Remove("cl");
                this.sum_menge["l"] = liter;
            }

            if (this.sum_menge.ContainsKey("g") && this.sum_menge.ContainsKey("kg"))
            {
                decimal g   = this.sum_menge["g"];
                decimal kg  = this.sum_menge["kg"];

                decimal gewicht = kg + g / 1000;

                this.sum_menge.Remove("g");
                this.sum_menge["kg"] = gewicht;
            }

        }
    }
}

File: app/src/main/java/de/stryi/vorratsuebersicht/tools/StockStatistic.kt
import android.annotation.SuppressLint
import android.app.Activity
import de.stryi.vorratsuebersicht.R
import de.stryi.vorratsuebersicht.database.Records.ArticleInfo
import de.stryi.vorratsuebersicht.tools.Tools
import java.text.NumberFormat

class StockStatistic {

    internal var count = 0
    internal var quantity = 0.0     // Double statt decimal
    private val sumMenge = mutableMapOf<String, Double>()
    private var sumWarnung = 0.0
    private var sumAbgelaufen = 0.0
    private var sumKcal = 0.0
    private var sumPrice = 0.0

    /*
    fun addWarningLevel1(quantity: Double) {
        sumWarnung += quantity
    }

    fun addWarningLevel2(quantity: Double) {
        sumAbgelaufen += quantity
    }
    */

    private fun addUnitQuantity(unit: String?, size: Double?, quantity: Double) {
        if (size == null || size <= 0.0)
            return

        val key = unit ?: ""
        sumMenge[key] = (sumMenge[key] ?: 0.0) + size * quantity
    }

    private fun addCalorie(quantity: Double, calorie: Int?) {
        if (calorie == null || calorie <= 0)
            return

        sumKcal += quantity * calorie
    }

    fun addStorageItem(item: ArticleInfo) {
        count++
        quantity += item.storageQuantity
        addUnitQuantity(item.unit, item.size, item.storageQuantity)
        addCalorie(item.storageQuantity, item.calorie)
        addCosts(item.storageQuantity, item.price)
    }

    private fun addCosts(quantity: Double, price: Double?) {
        if (price != null) {
            sumPrice += quantity * price
        }
    }

    fun getText(storageItemQuantityList: List<ArticleInfo>, activity: Activity):String {

        for (item in storageItemQuantityList) {

            val itemList = item.storageArticleList
            for (item in itemList)
            {
                if (item.warningLevel == 1)
                {
                    sumWarnung += item.quantity
                }

                if (item.warningLevel == 2)
                {
                    sumAbgelaufen += item.quantity
                }
            }

            addStorageItem(item)
        }

        return getStatistic(activity)
    }

    @SuppressLint("StringFormatMatches", "StringFormatInvalid")
    fun getStatistic(activity: Activity): String {
        val res = activity.resources
        var status: String

        if (count == 1)
            status = String.format(res.getString(R.string.StorageListSummary_Position), count)
        else
            status = String.format(res.getString(R.string.StorageListSummary_Positions), Tools.formatNumber(count))

        val quantityText = res.getString(R.string.StorageListSummary_Quantity)

        status += ", " + String.format(quantityText, Tools.formatNumber(quantity))

        if (sumMenge.isNotEmpty()) {
            convertUnits()

            /*
            val mengeListe = sumMenge
                .filter { it.value > 0 }
                .entries.joinToString(", ") {
                    String.format(Locale.getDefault(), "%,.######f %s", it.value, it.key)
                }

            if (mengeListe.isNotEmpty()) {
                append(", ")
                append(
                    String.format(
                        res.getString(R.string.StorageListSummary_Amount),
                        mengeListe
                    )
                )
            }
            */

            if (sumWarnung > 0)
            {
                status += ", " + String.format(
                    res.getString(R.string.StorageListSummary_Warning),
                    Tools.formatNumber(sumWarnung))
            }

            if (sumAbgelaufen > 0)
            {
                status += ", " + String.format(
                    res.getString(R.string.StorageListSummary_Off),
                    Tools.formatNumber(sumAbgelaufen))
            }
        }

        if (sumKcal > 0)
            status += ", " + String.format(
                res.getString(R.string.StorageListSummary_Calories),
                Tools.formatNumber(sumKcal))

        if (sumPrice > 0)
            status += ", " + String.format(
                    res.getString(R.string.StorageListSummary_Value),
                    Tools.formatNumber(sumPrice),
                    NumberFormat.getCurrencyInstance().currency?.symbol ?: "")

        /*
        if (sumWarnung > 0)
            status = ", " + String.format(
                res.getString(R.string.StorageListSummary_Warning), Tools.formatNumber(sumWarnung))

        if (sumAbgelaufen > 0)
            status = ", " + String.format(
                res.getString(R.string.StorageListSummary_Off), Tools.formatNumber(sumAbgelaufen))
        */

        return status
    }

    private fun convertUnits() {
        if (sumMenge.containsKey("ml") && sumMenge.containsKey("l")) {
            val liter = (sumMenge["l"] ?: 0.0) + (sumMenge["ml"] ?: 0.0) / 1000
            sumMenge.remove("ml")
            sumMenge["l"] = liter
        }

        if (sumMenge.containsKey("cl") && sumMenge.containsKey("l")) {
            val liter = (sumMenge["l"] ?: 0.0) + (sumMenge["cl"] ?: 0.0) / 100
            sumMenge.remove("cl")
            sumMenge["l"] = liter
        }

        if (sumMenge.containsKey("g") && sumMenge.containsKey("kg")) {
            val gewicht = (sumMenge["kg"] ?: 0.0) + (sumMenge["g"] ?: 0.0) / 1000
            sumMenge.remove("g")
            sumMenge["kg"] = gewicht
        }
    }
}
